Functions and instances from base-4.7 and base-4.8

RyanGlScott opened the pull request:
I noticed there are a lot of new introduced functions and instances from `base-4.7` and `base-4.8` (at least, I noticed what was put into the [changelog](http://hackage.haskell.org/package/base-4.8.0.0/changelog)), so I took a shot at implementing them. The extent of the changes are:

* Added `Eq`, `Ord`, `Read`, and `Show` instances for data types in `GHC.Generics`
* Added `Eq`, `Ord`, `Read`, and `Show` instances for `Const`
* Added `Read` and `Show` instances for `Down`
* Added `Eq`, `Ord`, `Read`, and `Show` instances for `ZipList`
* Added `Data` and `IsList` instances for `Version`
* `Bits` instance for `Bool`
* `Generic` instance for `All`, `Any`, `Const`, `Dual`, `Endo`, `First`, `Last`, `Product`, `Sum`, `WrappedArrow`, `WrappedMonad`, and `ZipList`
* `Generic1` instance for `Const`, `Dual`, `First`, `Last`, `Product`, `Sum`, `WrappedArrow`, `WrappedMonad`, and `ZipList`
* `Functor` instance for `ArgOrder`, `OptDescr`, and `ArgDescr`
* `Num` instance for `Sum` and `Product`
* `Storable` instance for `Complex` and `Ratio`
* Added `withMVarMasked` function to `Control.Concurrent.MVar.Compat`
* Added `(<$!>)` function to `Control.Monad.Compat`
* `(&)` function to `Data.Function.Compat`
* `($>)` and `void` functions to `Data.Functor.Compat`
* `dropWhileEnd`, `isSubsequenceOf`, `sortOn`, and `uncons` functions to `Data.List.Compat`
* `makeVersion` function to `Data.Version.Compat`
* `traceId`, `traceShowId`, `traceM`, and `traceShowM` functions to `Debug.Trace.Compat`
* `calloc` and `callocBytes` functions to `Foreign.Marshal.Alloc.Compat`
* `callocArray` and `callocArray0` functions to `Foreign.Marshal.Array.Compat`

I avoided changes which would involve reimplementing data types (e.g., `Alt`, which was introduced in `Data.Monoid` in `base-4.8.0.0`), since I'm not sure if you intend for `base-compat` to do that (like `transformers-compat` does).

sol commented:
@RyanGlScott Hey, this looks like a massive contribution.  I'll review it ASAP.

Regarding data types, I think it's fine to add new data types.  But I think if an existing data type changed, we should not reimplement it, as I think this does not compose well.

sol commented on base-compat.cabal:
```diff
@@ -35,18 +35,35 @@ library
+  if impl(ghc >= 7.2)
+    exposed-modules:
+      GHC.Generics.Compat
+    if impl(ghc < 7.6)
```
Would it be harmful to remove this condition?

RyanGlScott replied:
The only catch is that the `GHC.Generics` module is only defined on GHC 7.2 and later, so I'd have to add some more CPP pragmas to ensure that `GHC.Generics.Compat` doesn't export anything on GHC 7.0.

sol replied:
I was talking only about the second condition, then one for `ghc-prim`.

sol replied:
But as I mentioned elsewhere, I think it's good now :)

sol commented on src/Control/Applicative/Compat.hs:
```diff
@@ -1,3 +1,6 @@
+#if !MIN_VERSION_base(4,8,0)
+{-# LANGUAGE StandaloneDeriving #-}
+#endif
```
Again, if we can remove the CPP conditional here without adverse effects, I would vote for that.

sol commented on src/Control/Applicative/Compat.hs:
```diff
@@ -19,4 +23,22 @@ instance Monoid a => Monoid (Const a b) where
+
+instance Show a => Show (Const a b) where
+    showsPrec d (Const x) = showParen (d > 10) $
+                            showString "Const " . showsPrec 11 x
```
What is the reason we can't derive those instances, e.g.:

```haskell
deriving instance Read a => Read (Const a b)
deriving instance Show a => Show (Const a b)
```

RyanGlScott replied:
Weirdly enough, those definitions are [what `base` currently uses](http://hackage.haskell.org/package/base-4.8.0.0/docs/src/Control-Applicative.html#Const). If they were derived, then `show (Const "a")` would result in `"Const {getConst = \"a\"}"` instead of `"Const \"a\""`.

sol replied:
Ok, I think it makes sense to go for the handwritten instances then.

sol commented on src/Control/Monad/Compat.hs:
```diff
@@ -1,15 +1,33 @@
+f <$!> m = do
+  x <- m
+  let z = f x
+  z `seq` return z
```
I think we need a
```haskell
infixl 4 <$!>
```
here.

sol replied:
Good now, thanks!

sol commented on src/Data/Bool/Compat.hs:
```diff
@@ -1,3 +1,6 @@
+#if MIN_VERSION_base(4,6,0) && !MIN_VERSION_base(4,7,0)
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+#endif
```
Again, I would remove the CPP conditional

RyanGlScott replied:
Works for me. I got a bit carried away with the pragmas (in case you ever decided to run `hlint` on this and wanted to avoid unused language extensions). I'll fix all instances of this in other files.

sol commented on src/Data/Bool/Compat.hs:
```diff
@@ -12,6 +15,41 @@ module Data.Bool.Compat (
+import Data.Bits (Bits(..))
+import Data.Eq ((/=))
+
+instance Bits Bool where
```
What about moving this to `Data.Bits.Compat`?  My reasoning is:

 * In `base` the instance is defined in `Data.Bits`
 * To use the instance a user has to import `Data.Bits`/`Data.Bits.Compat` anyway.  Moving the instance to `Data.Bits.Compat` allows the user to use the instance with only one import instead of two.

RyanGlScott replied:
Sure, works for me. It also hit me that there are some orphan instances that are available in one module and not the other (e.g., there's a `Generic (Const a b)` instance in `GHC.Generics.Compat` but not in `Control.Applicative.Compat`), so I'll fix that too.

sol replied:
Yeah, that sounds good.  We may eventually move all orphans into one central module that we import from all the other modules.

sol commented on src/Data/Monoid/Compat.hs:
```diff
@@ -1,3 +1,7 @@
+#if !MIN_VERSION_base(4,7,0)
+{-# LANGUAGE GeneralizedNewtypeDeriving, StandaloneDeriving #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+#endif
```
Again, I would tend to remove the CPP conditional.

sol commented on src/Data/Ord/Compat.hs:
```diff
@@ -0,0 +1,15 @@
+#if MIN_VERSION_base(4,6,0) && !MIN_VERSION_base(4,7,0)
+{-# LANGUAGE StandaloneDeriving #-}
+#endif
```
Same here regarding CPP.

sol commented on src/Data/Version/Compat.hs:
```diff
@@ -0,0 +1,52 @@
+#if !MIN_VERSION_base(4,8,0)
+import Prelude.Compat
+
+# if __GLASGOW_HASKELL__ >= 708
```
I think it would also be fine to constrain by `base` version here.  But I guess both are fine.

sol commented on src/Data/Version/Compat.hs:
```diff
@@ -0,0 +1,52 @@
+  gunfold k z c = case constrIndex c of
+                    1 -> k (k (z Version))
+                    _ -> error "Data.Data.gunfold(Version)"
+  dataTypeOf _  = versionDataType
```
I'm puzzled for what reason this instance is hand written instead of derived.  I see it's the same in `base`.  Do you have any insight here?

sol replied:
If deriving gives us the same observable behavior, then I'm strongly in favor of deriving.

RyanGlScott replied:
No idea why it's hand-written. All I can find is [this thread](http://permalink.gmane.org/gmane.comp.lang.haskell.libraries/20269) from a mailing list.

RyanGlScott replied:
A quick experiment using GHCi shows that this is the result of autoderiving a `Data` instance:

```haskell
λ> data Version = Version { versionBranch :: [Int], versionTags :: [String] } deriving Data

==================== Derived instances ====================
Derived instances:
  instance Data.Data.Data Ghci1.Version where
    Data.Data.gfoldl k_a2dy z_a2dz (Ghci1.Version a1_a2dA a2_a2dB)
      = ((z_a2dz Ghci1.Version `k_a2dy` a1_a2dA) `k_a2dy` a2_a2dB)
    Data.Data.gunfold k_a2dC z_a2dD _
      = k_a2dC (k_a2dC (z_a2dD Ghci1.Version))
    Data.Data.toConstr (Ghci1.Version _ _) = Ghci1.$crvY
    Data.Data.dataTypeOf _ = Ghci1.$trvc
  
  Ghci1.$trvc :: Data.Data.DataType
  Ghci1.$crvY :: Data.Data.Constr
  Ghci1.$trvc = Data.Data.mkDataType "Version" [Ghci1.$crvY]
  Ghci1.$crvY
    = Data.Data.mkConstr
        Ghci1.$trvc
        "Version"
        ["versionBranch", "versionTags"]
        Data.Data.Prefix
```

which, as far as I can tell, is equivalent to the hand-rolled definition (there's no case that leads to an `error`, but I don't think that could ever happen anyway since `Version` has exactly one constructor).

sol replied:
Nice! :)

sol commented on src/Foreign/Marshal/Alloc/Compat.hs:
```diff
@@ -0,0 +1,44 @@
+#if !MIN_VERSION_base(4,8,0)
+{-# LANGUAGE ForeignFunctionInterface #-}
+#endif
```
Again, same thing regarding CPP.

sol commented on src/Foreign/Marshal/Alloc/Compat.hs:
```diff
@@ -0,0 +1,44 @@
+#if !MIN_VERSION_base(4,8,0)
+{-# LANGUAGE ForeignFunctionInterface #-}
+#endif
+module Foreign.Marshal.Alloc.Compat (
```
Now that we have `Foreign.Marshal` re-exports, both `Foreign.Marshal.Alloc` and `Foreign.Marshal.Array`.  So now that we have `Compat` version of those modules, I think it would be nice to have `Foreign.Marshal.Compat` too, that re-exports the `Compat` versions.  But we could do that as a separate change.

sol replied:
And this propagates up one more level to `Foreign`, which re-exports `Foreign.Marshal`.

sol commented on test/Data/Version/CompatSpec.hs:
```diff
@@ -0,0 +1,28 @@
+  describe "IsList Version instance" $
+    it "creates a Version from an Int list" $
+      [1,2,3] `shouldBe` makeVersion [1,2,3]
+#endif
```
Newline at the end of the file please ;)

sol commented on src/Foreign/Marshal/Alloc/Compat.hs:
```diff
@@ -0,0 +1,42 @@
+--
+{-# INLINE calloc #-}
+calloc :: Storable a => IO (Ptr a)
+calloc = doCalloc undefined
```
Some basic test for some of the functions in here would be nice, so that we at least have some evidence that the `foreign import` is correct.

sol commented on src/GHC/Generics/Compat.hs:
```diff
@@ -0,0 +1,115 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module GHC.Generics.Compat (
+-- GHC.Generics is only available on GHC 7.2 and later, so we can't export anything
+-- if we're using a GHC prior to 7.2
```
Yes, I think an empty module stub is fine for GHC < 7.2 :)

sol commented on test/Control/Applicative/CompatSpec.hs:
```diff
@@ -30,6 +30,24 @@ spec = do
+  describe "Show (Const a b)" $
+    it "allows a Const value to be converted to a String" $
+      show (Const 1 :: Const Int Int) `shouldBe` "Const 1"
+
```
I'm on the fence for writing specs for derived code.  It's nice to ensure that instances behave the same across multiple versions of GHC, but on the other hand we actually test that GHC's deriving mechanism works here.

If I would write it, I'm still undecided whether I want to have specs in cases like that or not.

sol commented on src/Data/Bits/Compat.hs:
```diff
@@ -0,0 +1,38 @@
+import Data.Bits as Base
+
+-- These instances are only valid if Bits isn't a subclass of Num (as Bool is
+-- not a Num instance), which is only true as of base-4.6.0.0 and later.
```
This is unfortunate, but I guess there is nothing we can do about it :(

sol commented on test/Data/Ord/CompatSpec.hs:
```diff
@@ -0,0 +1,35 @@
+main = hspec spec
+
+spec :: Spec
+#if MIN_VERSION_base(4,6,0)
```
I'm in favor of removing the conditional and adding

```haskell
newtype Down a = Down a 
```

for bale < 4.6.0.

RyanGlScott replied:
I am too, although I think I'll wait and do that in a separate pull request, since there'd be quite a few data types I'd want to redefine (off the top of my head, `Bifunctor`, `Identity`, `Alt`, `Void`, `Nat`, etc.).

sol replied:
+1

sol commented:
@RyanGlScott This looks awesome.  Let's sort out the things I commented on and get this on master.

My only complaint is that it's a quite big PR.  For future contributions it may make sense to break it down to make reviewing easier.

RyanGlScott commented:
Sorry for the code dump. I figured I should "limit" this PR to just the changes from `base-4.7` and `base-4.8`, but that turned out to be a lot more than I had imagined!

I intentionally left out:

* Anything which redefined data types (e.g., `Alt` in `Data.Monoid`, `Down` in `Data.Ord`) or uses a data type previously defined in another package (e.g., `Bifunctor`, `Nat`, `Void`)
* Changes in earlier versions of `base`. The changelog only goes back to 4.7.0.0, so any changes prior to that are harder to find.

Also, I went ahead and incorporated your suggestions. Please review and let me know if I goofed anywhere (I'm trying to push this all in before I call it a night). In particular, review the `Foreign`-related code. I tried to add extra `.Compat` modules that do nothing but reexport other `Foreign` modules, and I added a test for `calloc`, but it may not be comprehensive enough depending on your code coverage philosophy.

Thanks for taking the time to look over this. Given how much pain changes to `base` can cause, I feel like `base-compat` is an extremely important package, so I'm happy to change it until I get it right.

sol commented on src/Foreign/Marshal/Compat.hs:
```diff
@@ -0,0 +1,9 @@
+import Foreign.Marshal as Base
+
+import Foreign.Marshal.Alloc.Compat as Alloc
+import Foreign.Marshal.Array.Compat as Array
```
Looks good :)

sol replied:
Except for the missing newline ;)

sol commented on src/Foreign/Compat.hs:
```diff
@@ -0,0 +1,11 @@
+
+import Data.Bits.Compat as Bits
+import Foreign.Marshal.Compat as Marshal
+import Foreign.Storable.Compat as Storable
```
Looks good to me :)

sol commented on test/Foreign/Marshal/Alloc/CompatSpec.hs:
```diff
@@ -0,0 +1,27 @@
+        free intPtr1
+        free intPtr2
+        return res
+        
```
trailing spaces

sol commented on test/Foreign/Marshal/Alloc/CompatSpec.hs:
```diff
@@ -0,0 +1,27 @@
+        res <- val1 `shouldBe` val2
+        free intPtr1
+        free intPtr2
+        return res
```
`res` is just `()` here

sol commented on test/Foreign/Marshal/Alloc/CompatSpec.hs:
```diff
@@ -0,0 +1,27 @@
+spec = do
+  describe "calloc" $
+    it "allocates memory with bytes of value zero" $
+      property $ do
```
Maybe I'm missing something obvious here, this property is not parameterized.  Does it make any difference to run it once vs. 100 times?

RyanGlScott replied:
No, it certainly doesn't need to be run 100 times. Thanks for noticing that. :)

sol commented:
@RyanGlScott Except for my open question about the spec for `calloc`, it's good now.

sol commented:
@RyanGlScott I further simplified the test case for `calloc`, https://github.com/sol/base-compat/commit/e8e78a76d9b8bc0cd8be61f5ff8d5a27e6e6351e.  WDYT?

sol commented:
@RyanGlScott Please also feel free to add yourself to the `copyright` field in the cabal file and the LICENSE file.

sol commented:
On Hackage as 0.7.0.

RyanGlScott commented:
Looks great, thanks for uploading while I was asleep! I'll make some more PRs later to address the stuff I missed (like `Down`), in smaller chunks this time :)
